source/Jobbr.Server.WebAPI/Infrastructure/WebHost.cs

```diff
@@ -12,79 +13,81 @@ namespace Jobbr.Server.WebAPI.Infrastructure
 
         /// <summary>
         /// Initializes a new instance of the <see cref="WebHost"/> class.
         /// </summary>
-        public WebHost(IJobbrServiceProvider dependencyResolver, JobbrWebApiConfiguration configuration)
+        /// <param name="loggerFactory">The logger factory.</param>
+        /// <param name="container">The service injector container.</param>
+        /// <param name="configuration">Web API configuration.</param>
+        public WebHost(ILoggerFactory loggerFactory, Container container, JobbrWebApiConfiguration configuration)
         {
-            this.dependencyResolver = dependencyResolver;
-            this.configuration = configuration;
+            _logger = loggerFactory.CreateLogger<WebHost>();
+            _serviceCollection = container.GetCurrentRegistrations();
+            _configuration = configuration;
         }
 
         /// <summary>
-        /// Start the component
+        /// Start the web host.
         /// </summary>
         public void Start()
         {
-            this.AssertBackendAddressIsValid();
-            
-            var services = (ServiceProvider)ServicesFactory.Create();
-            var options = new StartOptions()
+            if (_webHost != null)
             {
-                Urls = { this.configuration.BackendAddress },
-                AppStartup = typeof(Startup).FullName
-            };
-            // Pass through the IJobbrServiceProvider to allow Startup-Classes to let them inject this dependency
-            services.Add(typeof(IJobbrServiceProvider), () => this.dependencyResolver);
+                throw new InvalidOperationException("The server has already been started.");
+            }
+
+            AssertBackendAddressIsValid();
 
-            var hostingStarter = services.GetService<IHostingStarter>();
-            this.webHost = hostingStarter.Start(options);
+            _webHost = new WebHostBuilder()
+                .UseKestrel()
+                .UseUrls(new Uri(_configuration.BackendAddress).GetLeftPart(UriPartial.Authority))
+                .ConfigureServices(services =>
+                {
+                    foreach (var instanceProducer in _serviceCollection)
+                    {
+                        services.Add(new ServiceDescriptor(instanceProducer.ServiceType, instanceProducer.GetInstance()));
+                    }
+                })
+                .UseStartup<Startup>()
+                .Build();
 
-            Logger.InfoFormat($"Started OWIN-Host for WebAPI at '{this.configuration.BackendAddress}'.");
+            _webHost.Start();
+
+            _logger.LogInformation("Started web host for WebAPI at '{backendAddress}'.", _configuration.BackendAddress);
         }
 
         public void Stop()
         {
-            Logger.InfoFormat("Stopping OWIN-Host for Web-Endpoints'");
-
-            this.webHost?.Dispose();
-
-            this.webHost = null;
+            Task.FromResult(_webHost.StopAsync());
```

Comment:
[Some say](https://stackoverflow.com/questions/9343594/how-to-call-asynchronous-method-from-synchronous-method-in-c) that there are better methods than `Task.FromResult`, but I'm not sure. Always seems like a topic nobody has a good answer for

Reply:
You are right, there are multiple solutions out there, I still don't know if it's the best one but I think it should do.